p2p/host/basic/basic_host.go

```diff
@@ -143,17 +150,30 @@ func NewHost(net inet.Network, opts *HostOpts) *BasicHost {
 		h.cmgr = opts.ConnManager
 	}
 
+	var relayCtx context.Context
+	var relayCancel func()
+	if opts.Relay {
+		relayCtx, relayCancel = context.WithCancel(context.Background())
```

Comment:
It's loosely wired via the cancel funciton. When the basic host gets closed, it will call cancel.

Regardless, I agree that ideally this context should come from the outside and be tied to the lifetime of the host. It would take an interface change to pass it as a `NewHost` argument though, so that's why I didn't do that initially.

But we _are_ changing the interface already for the error return. It's not widely used so we might as well change it further to take a context argument. 
The legacy interface can pass in a context.Background().